experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+
+        compiled_f = compile_fx_inner(fnfx, used_args)
+
+    def inductor_fn(*args):
+        with torch.no_grad():
+            used_args = prepare_args(real_attr_vals, unused_input_idx, args)
+
+            # if args are DTensors, get a local shard. But we also need to check
+            # that their placements match the intended placement.
+            for i in range(len(used_args)):
+                arg = used_args[i]
+                if isinstance(arg, DTensor):
+                    assert (
+                        arg.placements == schemas[i].placements and
+                        arg.device_mesh == schemas[i].mesh
+                    )
+                    used_args[i] = arg.to_local()
+
+            outputs = compiled_f(used_args)
+
+            return outputs if output_schemas is None else [
+                output if schema is None else DTensor.from_local(
+                    local_tensor=output,
+                    device_mesh=schema.mesh,
+                    placements=schema.placements,
+                    run_check=False,
+                )
+                for output, schema in zip(outputs, output_schemas)
+            ]
+
+
+    return inductor_fn
+
+
+# =====================================
+# 1. Minimal SGD example, no nn.Module
+# =====================================
+
+def inplace_test(x, p):
+    p.grad = None  # zero_grad
+    loss = (torch.sigmoid(torch.matmul(x, p)) ** 2).sum()
+    # pg, = torch.autograd.grad(loss, p)
+    loss.backward()
+    with torch.no_grad():
+        p -= 0.1 * p.grad
+    return (loss, )
+
+
+def main_simple():
+    fake_mode = FakeTensorMode()
+    torch.set_default_device('cuda')
+
+    if DIST:
+        mesh = DeviceMesh('cuda', range(torch.distributed.get_world_size()))
+        default_shard_schema = Schema(mesh, [Replicate()])
+
+    # with fake_mode:
+    with fake_mode:
+        x_fake = torch.randn(10, 2, device='cuda')
+        p_fake = torch.randn(2, 1, requires_grad=True, device='cuda')
+
+        if DIST:
+            mark_placements(x_fake, [Shard(0)])
+            mark_placements(p_fake, [Replicate()])
+
+            # x_shard = get_example_shard(x, mesh)
+            # dprint(f'Input shape: {x.shape}; shard shape: {x_shard.shape}')
+        else:
+            # x_shard = x
+            default_shard_schema = None
+
+        compiled_f = make_inductor_fn(
+            inplace_test,
+            [x_fake, p_fake],
+            default_shard_schema=default_shard_schema,
+        )
+
+    x = torch.randn(10, 2, device='cuda')
+    p = torch.randn(2, 1, requires_grad=True, device='cuda')
+
+    if DIST:
+        x = distribute_tensor(x, mesh, [Shard(0)])
+        p = distribute_tensor(p, mesh, [Replicate()])
+
+    if DIST:
+        # make sure our parameter is the same across devices initially
+        # doing this after compile because during compilation we're running the training step a few times
+        # TODO: fix this. we should run compilation with fake tensors.
```

Comment:
I am guessing this is not true anymore since you got FakeTensors to work above.

Reply:
right, in the new version, i am calling 

mc.linear.reset_parameters()

which will act inplace on DTensors and will effectively initalize them in a sharded way. 
I had to implement a DTensor.uniform_ propagation rule in order to make this work -- such that it calls uniform_ on each shard of the parameter. This is nice, however it is not invariant to resharding (we shouold ideally have a version of random ops that takes a shard-invariant random seed, as previously discussed with Wanchao, if we want to preserve invariance of the op over resharding).